src/launchHandler.py

```diff
@@ -18,81 +22,39 @@ def __init__(self):
 def launchFile(package, fileName):
     command = "roslaunch  {0} {1}".format(package, fileName)
 
-    p = subprocess.Popen(command, shell=True)
+    p = subprocess.Popen(command, shell=True, preexec_fn=os.setsid)
+
     launchMsg = LaunchMsg()
     #Sleep to make sure the launch command has time to fail if there's an error
     sleep(1)
     state = p.poll()
     launchMsg.fileName = fileName
     if state is None:
         launchMsg.message = fileName + " was launched"
-        launchedFiles[fileName] = package
+        launchedFiles[fileName] = LaunchFile(package, fileName, p.pid)
         launchMsg.isLaunched = True
     else:
         launchMsg.message = fileName + " was not launched"
         launchMsg.isLaunched = False
     return launchMsg
 
-def killLaunchFile(package, fileName):
-    isLaunched = False
-    isLaunched = recursivekillLaunchFile(package, fileName)
+def killLaunchFile(fileName):
     launchMsg = LaunchMsg()
     launchMsg.fileName = fileName
-    if(not isLaunched):
-        launchedFiles.pop(fileName)
-        launchMsg.message = "Nodes in " + fileName + " were killed"
+    if fileName in launchedFiles:
+        pid = launchedFiles[fileName].pid
+        os.killpg(os.getpgid(pid), signal.SIGINT)
```

Comment:
Is it necessary to use SIGINT (maybe ROS ask that we use SIGINT for killing nodes) ? SIGINT is meant to interrupt a process, similar to what we do when doing ctrl-C in a terminal. But doesn't really mean terminating a process, SIGTERM does. Also, both SIGINT and SIGTERM can be repurposed, like we did in the e-stop node. If we really want to **KILL** the node, we should use SIGKILL, which cannot be repurposed and will kill the node **non-gracefully**. (https://stackoverflow.com/questions/4042201/how-does-sigint-relate-to-the-other-termination-signals-such-as-sigterm-sigquit)

Reply:
My goal was to keep the launch handler procedure as close as the procedure we use when using the terminal, so SIGINT was used to simulate a CTRL+C. But maybe SIGKILL could be better and would assure that the process will be killed. My only concern is if we use SIGKILL will we pass in a node's shutdown callback?

Reply:
No, we won't pass in a shutdown/clean-up function from a node. That might be alright to keep SIGINT in that case. I wouldn't be surprise that ROS node have some kind of shutdown procedure that need to be down before shutting down. Thanks for clearing this up with me :).

Reply:
During testing when SIGINT wasn't working ROS automatically tried SIGTERM then SIGKILL to shutdown the node.